src/type_resolve_helpers.ts

```diff
@@ -4,15 +4,316 @@ import { warn } from './logger';
 import { PropTree, IPropDesc } from './PropTree';
 import { type } from 'os';
 
-const rgxObjectTokenize = /(<|>|,)/;
+const rgxObjectTokenize = /(<|>|,|\(|\)|\|)/;
 const rgxCommaAll = /,/g;
 const rgxParensAll = /\(|\)/g;
 
 const anyTypeNode = ts.createKeywordTypeNode(ts.SyntaxKind.AnyKeyword);
 const strTypeNode = ts.createKeywordTypeNode(ts.SyntaxKind.StringKeyword);
 
-const anyGeneric: IGenericType = { kind: 'type', name: 'any', resolved: anyTypeNode };
-const strGeneric: IGenericType = { kind: 'type', name: 'string', resolved: strTypeNode };
+enum tsNodeTypes {
+    GENERIC,    // Foo.<X,Y>    has types for children
+    UNION,      // (a|b|c)      has types for children
+    FUNCTION,   // function()   has arguments for children
+    TUPLE,      // [a,b]        has types for children
+    TYPE        // string, X    has no children
+}
+class StringTreeNode {
+    name: string;
+    children: StringTreeNode[];
+    type: tsNodeTypes;
+    parent: StringTreeNode | null;
+    constructor(name: string, type: tsNodeTypes, parent: StringTreeNode | null)
+    {
+        this.name = name;
+        this.type = type;
+        this.parent = parent;
```

Comment:
This can be simplified by adding `public` to the constructor args and removing the assignments here and the decls above. Similarly the `= []` can be done in the children decl above.